storage/replicate_queue.go

```diff
@@ -113,40 +89,50 @@ func (rq replicateQueue) process(now proto.Timestamp, repl *Replica) error {
 	}
 
 	desc := repl.Desc()
-	delta := rq.replicaDelta(zone, repl, desc)
-	if delta == 0 {
-		// Something changed between shouldQueue and process.
+	action, _ := rq.allocator.computeAction(zone, desc)
+	if action == aaNoop {
+		// No action to take, return without re-queueing.
 		return nil
 	}
 
-	// TODO(bdarnell): handle non-homogenous ReplicaAttrs.
-	if delta > 0 {
-		// Allow constraints to be relaxed if necessary.
-		newReplica, err := rq.allocator.allocateTarget(zone.ReplicaAttrs[0], desc.Replicas, true, nil)
+	// Avoid taking action if the range has too many dead replicas to make
+	// quorum.
+	deadReplicas := rq.allocator.findDeadReplicas(desc)
```

Comment:
I'm not sure if you need the special case for `aaNoop` anyway -- if the range doesn't have a quorum then you should always get `aaRemoveDead`. There could be a race since you're calling `findDeadReplicas` twice, but I think it's fine if we return an error for `aaNoop` in those rare cases.